services/graphite/service.go

```diff
@@ -1,11 +1,305 @@
 package graphite
 
-import "net"
+import (
+	"bufio"
+	"fmt"
+	"log"
+	"math"
+	"net"
+	"os"
+	"strconv"
+	"strings"
+	"sync"
+	"time"
 
-type Service struct{}
+	"github.com/influxdb/influxdb/cluster"
+	"github.com/influxdb/influxdb/tsdb"
+)
 
-func NewService(c Config) *Service { return &Service{} }
+const (
+	udpBufferSize = 65536
+)
 
-func (s *Service) Open() error    { return nil }
-func (s *Service) Close() error   { return nil }
-func (s *Service) Addr() net.Addr { return nil }
+type Service struct {
+	bindAddress      string
+	database         string
+	protocol         string
+	batchSize        int
+	batchTimeout     time.Duration
+	consistencyLevel cluster.ConsistencyLevel
+
+	parser *Parser
+
+	logger *log.Logger
+
+	host string // Race free access to the bound address.
+	wg   sync.WaitGroup
+	done chan struct{}
+
+	PointsWriter interface {
+		WritePoints(p *cluster.WritePointsRequest) error
+	}
+}
+
+// NewService returns an instance of the Graphite service.
+func NewService(c Config) *Service {
+	s := Service{
+		bindAddress:      c.BindAddress,
+		database:         c.Database,
+		protocol:         c.Protocol,
+		batchSize:        c.BatchSize,
+		batchTimeout:     time.Duration(c.BatchTimeout),
+		consistencyLevel: c.ConsistencyAsEnum(),
+		logger:           log.New(os.Stderr, "[graphite] ", log.LstdFlags),
+		done:             make(chan struct{}),
+	}
+
+	parser := NewParser()
+	parser.Separator = c.NameSeparator
+	parser.LastEnabled = c.LastEnabled()
+	s.parser = parser
+
+	return &s
+}
+
+// Open starts the Graphite input processing data.
+func (s *Service) Open() error {
+	var addr net.Addr
+	var err error
+
+	if strings.ToLower(s.protocol) == "tcp" {
+		addr, err = s.openTCPServer()
+	} else if strings.ToLower(s.protocol) == "udp" {
+		addr, err = s.openUDPServer()
+	} else {
+		return fmt.Errorf("unrecognized Graphite input protocol %s", s.protocol)
+	}
+
+	if err != nil {
+		return err
+	}
+
+	s.host = addr.String()
+	s.logger.Printf("%s Graphite input opened on %s", s.protocol, s.host)
+	return nil
+}
+
+// Close stops all data processing on the Graphite input.
+func (s *Service) Close() error {
+	close(s.done)
+	s.wg.Wait()
+	return nil
+}
+
+func (s *Service) Host() string {
+	return s.host
+}
+
+// openTCPServer opens the Graphite input in TCP mode and starts processing data.
+func (s *Service) openTCPServer() (net.Addr, error) {
+	ln, err := net.Listen("tcp", s.bindAddress)
+	if err != nil {
+		return nil, err
+	}
```

Comment:
Does `ln` get closed anywhere? It seems like you'd need to set it as `Service.ln` so that `Service.Close()` can close it.